File: Interview_Prep_App/native_java/retrofit/src/main/java/com/dreams/interviewprepapp/ui/main/MainActivity.java
package com.dreams.interviewprepapp.ui.main;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;

import com.dreams.interviewprepapp.R;
import com.dreams.interviewprepapp.repositories.Tc2rGithubRepository;
import com.dreams.interviewprepapp.repositories.models.response.Answer;
import com.dreams.interviewprepapp.repositories.models.response.AnswersResponse;
import com.dreams.interviewprepapp.repositories.models.response.Question;
import com.dreams.interviewprepapp.repositories.models.response.QuestionsResponse;
import com.dreams.interviewprepapp.ui.questions.DialogFragment;
import com.dreams.interviewprepapp.ui.questions.QuestionFragment;
import com.dreams.interviewprepapp.ui.score.ScoreActivity;
import com.dreams.interviewprepapp.util.OnFragmentInteractionListener;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Random;
import java.util.concurrent.Callable;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.tc2r.sharedresources.R.*;

public class MainActivity extends AppCompatActivity implements OnFragmentInteractionListener {
    // Static Variables
    private final static int QUIZ_SIZE = 10;

    // UI Variables
    private TextView titleTv, scoreTv;


    // Variables
    private Tc2rGithubRepository tc2rGithubRepository;
    private ArrayList<Question> quizList;
    private ArrayList<Question> testList;
    private ArrayList<Answer> answersList;
    private Random random;
    private int currentQuestion = 0;
    private int numOfCorrect = 0;
    private int scorePer = 0;
    private double pointPerQ, score;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tc2rGithubRepository = new Tc2rGithubRepository();

        // Initialize and assignments
        titleTv = findViewById(R.id.title_tv);
        scoreTv = findViewById(R.id.score_tv);
        testList = new ArrayList<>();
        quizList = new ArrayList<>();
        answersList = new ArrayList<>();
        random = new Random();

        // get Wrong answers from server
        getAnswers();
    }

    public void getAnswers() {
        tc2rGithubRepository.getAnswersList.enqueue(new Callback<AnswersResponse>() {
            @Override
            public void onResponse(@NonNull Call<AnswersResponse> call,
                                   @NonNull Response<AnswersResponse> response) {
                int statusCode = response.code();
                AnswersResponse answers = response.body();
                if((answers != null ? answers.getAnswersList() : null) != null) {
                    answersList.addAll(answers.getAnswersList());
                    getQuestions();
                }
            }

            @Override
            public void onFailure(@NonNull Call<AnswersResponse> call, @NonNull Throwable t) {
                Log.e("Answers Error", t.toString());
            }
        });
    }

    public void getQuestions() {
        tc2rGithubRepository.getQuestionsList.enqueue(new Callback<QuestionsResponse>() {
            @Override
            public void onResponse(@NonNull Call<QuestionsResponse> call,
                                   @NonNull Response<QuestionsResponse> response) {
                QuestionsResponse questionsResponse = response.body();
                if((questionsResponse != null ? questionsResponse.getQuestionsList() : null) != null) {
                    quizList.addAll(questionsResponse.getQuestionsList());
                    createQuiz();
                }
            }

            @Override
            public void onFailure(@NonNull Call<QuestionsResponse> call, @NonNull Throwable t) {
                Log.e("Questions API Failure", t.toString());
            }
        });
    }

    private void createQuiz() {
        Log.i(" Size: ", "QuestionList is: " + quizList.size());
        Log.i(" Size: ", "AnswerList is: " + answersList.size());

        // set booleanArray to be same size as quizList
        boolean[] selectedQuestion = new boolean[quizList.size()];

        // set an int to a random number in the quizList
        int randNum = random.nextInt(quizList.size());
        int i = 0;

        // sets the score system for quiz.
        // TODO: 5/31/2017 maybe remove this and simply divide correct answers by total questions
        pointPerQ = 1.0 / QUIZ_SIZE;

        // prevent from having repeated questions in a single quiz
        while(i < QUIZ_SIZE) {
            // if boolean at randNum in selectedQuestion is false
            if(!selectedQuestion[randNum] && quizList
                    .get(randNum)
                    .getQuestionType()
                    .equals("multi")) {

                // Add position randNum to test list;
                testList.add(quizList.get(randNum));
                // set this question selected to true.
                selectedQuestion[randNum] = true;
                i++;
            } else {
                // if question already selected, change randNum;
                randNum = random.nextInt(quizList.size());
            }
        }

        // On first run, start quiz without updating score
        nextQuestion(false);
    }

    public String goToNextQuestion(boolean isAnswerCorrect) {
        nextQuestion(isAnswerCorrect);
        return "True";
    }

    public void showDetails(Answer answer, boolean isAnswerCorrect) {
        // Show a details dialog fragment on top of the current screen.
        DialogFragment myDialogFragment = new DialogFragment(answer, new Callable<String>() {
            public String call() {
                // after user dismisses the dialogfragment, move on to next question.
                return goToNextQuestion(isAnswerCorrect);
            }});
        myDialogFragment.show(getSupportFragmentManager(), "details");
    }



    public void nextQuestion(boolean correctAnswer) {
        // if previous question was answered correctly
        // update variables accordingly.
        if(correctAnswer) {
            numOfCorrect++;
            score += pointPerQ;
            scorePer = (int) (score * 100);
            scoreTv.setText(String.format(Locale.US, "%s%d", getString(string.score_display_text), scorePer));
        }

        // if quiz is not complete, continue quiz with new QuestionFragment
        if(currentQuestion < QUIZ_SIZE) {
            // Fragments and Model Variables
            QuestionFragment newFragment = QuestionFragment.newInstance(
                    testList.get(currentQuestion), answersList);
            currentQuestion++;
            titleTv.setText(String.format(Locale.US, "%s%d of %d",
                                          getString(string.question_display_text),
                                          currentQuestion, QUIZ_SIZE));
            // UI Variables
            LinearLayout fragContainer = findViewById(R.id.fragment_container);
            FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
            ft.replace(fragContainer.getId(), newFragment);
            ft.commit();
        } else {
            // Quiz is over, go to final page!
            // create intent
            Intent intent = new Intent(this, ScoreActivity.class);

            // add variables to send.
            intent.putExtra("scorePercentage", scorePer);
            intent.putExtra("quizSize", QUIZ_SIZE);
            intent.putExtra("numCorrect", numOfCorrect);

            // use intent.
            startActivity(intent);
        }
    }

    @Override
    public void fragmentInitialized() {

    }
}

File: Interview_Prep_App/native_kotlin/base/src/main/java/com/nudennie/interviewprepapp/ui/main/MainActivity.kt
package com.nudennie.interviewprepapp.ui.main

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import com.nudennie.interviewprepapp.R.id
import com.nudennie.interviewprepapp.R.layout
import com.nudennie.interviewprepapp.controllers.AnswersController
import com.nudennie.interviewprepapp.controllers.QuestionsController
import com.nudennie.interviewprepapp.repositories.models.response.Answer
import com.nudennie.interviewprepapp.repositories.models.response.Question
import com.nudennie.interviewprepapp.ui.questions.DialogFragment
import com.nudennie.interviewprepapp.ui.questions.QuestionFragment
import com.nudennie.interviewprepapp.ui.score.ScoreActivity
import com.nudennie.interviewprepapp.util.OnFragmentInteractionListener
import java.util.Random

class MainActivity : AppCompatActivity(), OnFragmentInteractionListener {
    companion object {
        // Static Variables
        private const val QUIZ_SIZE = 4
    }

    // UI Variables
    private var fragContainer: LinearLayout? = null
    private var titleTv: TextView? = null
    private var scoreTv: TextView? = null

    // Variables
    private lateinit var answersList: ArrayList<Answer>

    private var newFragment: QuestionFragment? = null
    private var quizList: ArrayList<Question>? = null
    private var testList: ArrayList<Question>? = null
    private var random: Random? = null
    private var currentQuestion = 0
    private var numOfCorrect = 0
    private var scorePer = 0
    private var pointPerQ = 0.0
    private var score = 0.0
    private var questionsController: QuestionsController? = null
    private var answersController: AnswersController? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_main)

        // Initialize and assignments
        titleTv = findViewById(id.title_tv)
        scoreTv = findViewById(id.score_tv)
        testList = ArrayList()
        quizList = ArrayList()
        answersList = ArrayList()
        random = Random()

        // get Wrong answers from server
        answers
        questions
        createQuiz()
    }

    private val answers: Unit
        get() {
            answersController = AnswersController()
            answersList = answersController!!.answers
        }
    private val questions: Unit
        get() {
            questionsController = QuestionsController()
            quizList = questionsController!!.questions
        }

    private fun createQuiz() {
        Log.i(" Size: ", "QuestionList is: " + quizList!!.size)
        Log.i(" Size: ", "AnswerList is: " + answersList.size)

        // set booleanArray to be same size as quizList
        val selectedQuestion = BooleanArray(quizList!!.size)

        // set an int to a random number in the quizList
        var randNum = random!!.nextInt(quizList!!.size)
        var i = 0

        // sets the score system for quiz.
        pointPerQ = 1.0 / QUIZ_SIZE
        while (i < QUIZ_SIZE) {
            // if boolean at randNum in selectedQuestion is false
            if (!selectedQuestion[randNum] && quizList!![randNum].questionType == "multi") {
                // Add position randNum to test list;
                testList!!.add(quizList!![randNum])
                // set this question selected to true.
                selectedQuestion[randNum] = true
                i++
            } else {
                // if question already selected, change randNum;
                randNum = random!!.nextInt(quizList!!.size)
            }
        }
        // On first run, start quiz without updating score
        nextQuestion(false)
    }

    fun showDetails(answer: Answer, isAnswerCorrect: Boolean) {
        // Show a details dialog fragment on top of the current screen.
        val detailsDialogFragment = DialogFragment(
            answer = answer,
            func = {
                // after user dismisses the dialogfragment, move on to next question.
                nextQuestion(isAnswerCorrect)
            }
        )
        detailsDialogFragment.show(supportFragmentManager, "details")
    }

    private fun nextQuestion(correctAnswer: Boolean) {
        // if previous question was answered correctly
        // update variables accordingly.
        if (correctAnswer) {
            numOfCorrect++
            score += pointPerQ
            scorePer = (score * 100).toInt()
            scoreTv!!.text = getString(com.tc2r.sharedresources.R.string.score_display_text) + scorePer.toString()
        }
        // if quiz is not complete, continue quiz with new QuestionFragment
        if (currentQuestion < QUIZ_SIZE) {
            newFragment = QuestionFragment.newInstance(testList!![currentQuestion], answersList)
            currentQuestion++
            titleTv!!.text = getString(com.tc2r.sharedresources.R.string.question_display_text) + currentQuestion.toString() + " of " + Integer.toString(QUIZ_SIZE)
            fragContainer = findViewById<LinearLayout>(id.fragment_container)
            val ft = supportFragmentManager.beginTransaction()
            ft.replace(fragContainer!!.id, newFragment as Fragment)
            ft.commit()
        } else {
            // Quiz is over, go to final page!
            // create intent
            val intent = Intent(this, ScoreActivity::class.java)

            // add variables to send.
            intent.putExtra("scorePercentage", scorePer)
            intent.putExtra("quizSize", QUIZ_SIZE)
            intent.putExtra("numCorrect", numOfCorrect)

            // use intent.
            startActivity(intent)
        }
    }

    override fun fragmentInitialized() {}
}
